### ISSUE TITLE
2.x - groupBy() Parallelization Not Working like in 1.0

### ISSUE THREAD

**Issue by thomasnield** [ReactiveX/RxJava#5073]
Hi David, 

While I was writing the Packt book, I think I might have found a problem with RxJava 2.0, unless this was an intended change and I'm missing something. When I do the parallelization pattern with `groupBy()` and a modulus Integer, it works fine in RxJava 1.0. Everything parallelizes across multiple threads as expected. 

```java 

import rx.Observable;
import rx.schedulers.Schedulers;
import java.time.LocalTime;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicInteger;

public class Launcher {
    public static void main(String[] args) {

       int coreCount = Runtime.getRuntime().availableProcessors();
       AtomicInteger assigner = new AtomicInteger(0);

       Observable.range(1,10)
               .groupBy(i -> assigner.incrementAndGet() % coreCount)
               .flatMap(grp -> grp.observeOn(Schedulers.io())
                       .map(i2 -> intenseCalculation(i2))
               )
               .subscribe(i -> System.out.println("Received " + i + " "
                       + LocalTime.now() + " on thread "
                       + Thread.currentThread().getName()));

       sleep(20000);
    }

    public static <T> T intenseCalculation(T value) {
        sleep(ThreadLocalRandom.current().nextInt(3000));
        return value;
    }
    public static void sleep(long millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}
```

**OUTPUT:**

```
Received 1 20:17:23.422 on thread RxIoScheduler-2
Received 3 20:17:23.840 on thread RxIoScheduler-4
Received 4 20:17:24.053 on thread RxIoScheduler-5
Received 9 20:17:24.273 on thread RxIoScheduler-2
Received 5 20:17:24.695 on thread RxIoScheduler-6
Received 8 20:17:24.950 on thread RxIoScheduler-9
Received 6 20:17:25.242 on thread RxIoScheduler-7
Received 7 20:17:25.251 on thread RxIoScheduler-8
Received 2 20:17:25.324 on thread RxIoScheduler-3
Received 10 20:17:27.532 on thread RxIoScheduler-3

```

But if I switch to RxJava 2.0, this no longer parallelizes. Everything stays strictly serial and only one thread is used. 

```java

import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;

import java.time.LocalTime;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicInteger;

public class Launcher {
    public static void main(String[] args) {

       int coreCount = Runtime.getRuntime().availableProcessors();
       AtomicInteger assigner = new AtomicInteger(0);

       Observable.range(1,10)
               .groupBy(i -> assigner.incrementAndGet() % coreCount)
               .flatMap(grp -> grp.observeOn(Schedulers.io())
                       .map(i2 -> intenseCalculation(i2))
               )
               .subscribe(i -> System.out.println("Received " + i + " "
                       + LocalTime.now() + " on thread "
                       + Thread.currentThread().getName()));

       sleep(20000);
    }

    public static <T> T intenseCalculation(T value) {
        sleep(ThreadLocalRandom.current().nextInt(3000));
        return value;
    }
    public static void sleep(long millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}
```

**OUTPUT:**

```
Received 1 20:21:03.321 on thread RxCachedThreadScheduler-1
Received 9 20:21:05.273 on thread RxCachedThreadScheduler-1
Received 2 20:21:05.559 on thread RxCachedThreadScheduler-1
Received 10 20:21:07.390 on thread RxCachedThreadScheduler-1
Received 3 20:21:10.055 on thread RxCachedThreadScheduler-1
Received 4 20:21:11.962 on thread RxCachedThreadScheduler-1
Received 5 20:21:12.850 on thread RxCachedThreadScheduler-1
Received 6 20:21:15.779 on thread RxCachedThreadScheduler-1
Received 7 20:21:17.927 on thread RxCachedThreadScheduler-1
Received 8 20:21:20.540 on thread RxCachedThreadScheduler-1
```

Is there something new that I'm supposed to do? Or is this a possible behavior regression?

**Comment by thomasnield** [comment]
Hmmm... this makes things more interesting. Using a `subscribeOn()` instead of an `observeOn()` seems to make parallelization work. I don't know if this was intended or not. 

```java

import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;

import java.time.LocalTime;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicInteger;

public class Launcher {
    public static void main(String[] args) {

       int coreCount = Runtime.getRuntime().availableProcessors();
       AtomicInteger assigner = new AtomicInteger(0);

       Observable.range(1,10)
               .groupBy(i -> assigner.incrementAndGet() % coreCount)
               .flatMap(grp -> grp.subscribeOn(Schedulers.io())
                       .map(i2 -> intenseCalculation(i2))
               )
               .subscribe(i -> System.out.println("Received " + i + " "
                       + LocalTime.now() + " on thread "
                       + Thread.currentThread().getName()));

       sleep(20000);
    }

    public static <T> T intenseCalculation(T value) {
        sleep(ThreadLocalRandom.current().nextInt(3000));
        return value;
    }
    public static void sleep(long millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}
```
**OUTPUT:**

```
Received 8 20:27:03.291 on thread RxCachedThreadScheduler-8
Received 6 20:27:03.446 on thread RxCachedThreadScheduler-6
Received 5 20:27:03.495 on thread RxCachedThreadScheduler-5
Received 4 20:27:03.681 on thread RxCachedThreadScheduler-4
Received 7 20:27:03.989 on thread RxCachedThreadScheduler-7
Received 2 20:27:04.797 on thread RxCachedThreadScheduler-2
Received 1 20:27:05.172 on thread RxCachedThreadScheduler-1
Received 9 20:27:05.327 on thread RxCachedThreadScheduler-1
Received 10 20:27:05.913 on thread RxCachedThreadScheduler-2
Received 3 20:27:05.957 on thread RxCachedThreadScheduler-3
```

**Comment by akarnokd** [comment]
Works for me on RxJava 2.0.5.

**Comment by thomasnield** [comment]
Ah, okay. I was using 2.0.3 where apparently there was an issue. It looks like it was resolved by 2.0.5. Sorry I didn't check that first.
